lib/src/kvs/bootstrap/archive.rs

```diff
@@ -0,0 +1,528 @@
+use std::time::Duration;
+
+use rand::Rng;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+
+use crate::err::BootstrapCause::{ChannelRecvError, ChannelSendError};
+use crate::err::ChannelVariant::{BootstrapArchive, BootstrapTxSupplier};
+use crate::err::Error;
+use crate::kvs::bootstrap::{TxRequestOneshot, TxResponseOneshot};
+use crate::kvs::{ds, BootstrapOperationResult};
+use crate::sql::Uuid;
+
+/// This task will read input live queries from a receiver in batches and
+/// archive them and finally send them to the output channel.
+/// The task terminates if there is an irrecoverable error or if the input
+/// channel has been closed (dropped, from previous task).
+pub(crate) async fn archive_live_queries(
+	tx_req: mpsc::Sender<TxRequestOneshot>,
+	node_id: Uuid,
+	mut scan_recv: mpsc::Receiver<BootstrapOperationResult>,
+	sender: mpsc::Sender<BootstrapOperationResult>,
+	batch_size: usize,
+	batch_latency: &Duration,
+) -> Result<(), Error> {
+	let mut msg: Vec<BootstrapOperationResult> = Vec::with_capacity(batch_size);
+	loop {
+		match tokio::time::timeout(*batch_latency, scan_recv.recv()).await {
+			Ok(Some(bor)) => {
+				let is_err = bor.1.is_some();
+				if is_err {
+					// send any errors further on, because we don't need to process them
+					// unless we can handle them. Currently we can't.
+					// if we error on send, then we bubble up because this shouldn't happen
+					sender.send(bor).await.map_err(|e| {
+						error!("Error sending error: {:?}", e);
+						Error::BootstrapError(ChannelSendError(BootstrapArchive))
+					})?;
+				} else {
+					msg.push(bor);
+					if msg.len() >= batch_size {
+						let results =
+							archive_live_query_batch(tx_req.clone(), node_id, &mut msg).await?;
+						for boresult in results {
+							sender.send(boresult).await.map_err(|e| {
+								error!("Error sending error: {:?}", e);
+								Error::BootstrapError(ChannelSendError(BootstrapArchive))
```

Comment:
Same here.

Reply:
Panicking

Reply:
Errors on sending or receiving on a channel are common and expected. We should handle them correctly. Panicking is not a good error handling strategy in production code for expected errors. Is there any reason we can't simply log a message and exit the loop? Are you thinking that `sender` channel should always be open to receive new messages?